required/latex-lab/latex-lab-toc-kernel-changes.dtx

```diff
@@ -0,0 +1,429 @@
+%
+% The followings contains various changes to kernel commands
+% needed for the tagging of table of contents and similar lists.
+%    \begin{macrocode}
+%<*package>
+%    \end{macrocode}
+% \section{General kernel and class changes}
+% 
+% Some of the changes can probably could go into the kernel. 
+% This are marked with [kernel?],
+% the other are marked with [latex-lab]. A few changes must be done in the classes.
+% They are marked as [class]. They are inserted with hooks here: this will break
+% if a non-standard class is used.
+%
+% \subsection{Providing the counter representation \cs{theHxx} generally}
+% [kernel?]
+%
+% The \cs{theHxx} representation allows to create a unique representation of a counter
+% that is for example used to create destination names. It will also be needed
+% to add |/Ref| keys to various tagging structures. It makes therefore sense
+% to provide it by default as soon as a counter is created. The |\@addtoreset| change
+% (taken also from hyperref) tries to ensure the counter stays unique if it is reset.
+%
+% At first we suppress the patches from hyperref:
+%    \begin{macrocode}
+\def\hyper@nopatch@counter{}
+%    \end{macrocode}
+%
+%    \begin{macrocode}
+\def\@definecounter#1{\expandafter\newcount\csname c@#1\endcsname
+     \setcounter{#1}\z@
+     \global\expandafter\let\csname cl@#1\endcsname\@empty
+     \@addtoreset{#1}{@ckpt}%
+     \global\expandafter\let\csname p@#1\endcsname\@empty
+     \expandafter\gdef\csname theH#1\endcsname{\the\value{#1}}%new
+     \expandafter
+     \gdef\csname the#1\expandafter\endcsname\expandafter
+          {\expandafter\@arabic\csname c@#1\endcsname}}
+\def\@addtoreset#1#2{\expandafter\@cons\csname cl@#2\endcsname {{#1}}%
+  \expandafter\gdef\csname theH#1\endcsname{\csname theH#2\endcsname.\the\value{#1}}%
+}
+%    \end{macrocode}
+% The following counters are defined in the kernel
+%    \begin{macrocode}
+\gdef\theHenumi{\the\value{enumi}}
+\gdef\theHenumii{\the\value{enumi}}
+\gdef\theHenumiii{\the\value{enumi}}
+\gdef\theHenumiv{\the\value{enumi}}
+%    \end{macrocode}
+%
+% \subsection{Providing and updating \cs{@currentHref}}
+% [kernel?]
+%
+%  \cs{@currentHref} contains the current unique representation of a counter.
+%  It is useful also without hyperref as it allows to connect with a |/Ref|
+%  key a toc entry with the heading it refers too. So \cs{refstepcounter}
+%  is extended to update the command. This is done at the begin so that the
+%  hyperref code can still adapt it later (using the cmd/refstepcounter/after hook
+%  would be too late, and could also break with cleverref.)
+%
+%    \begin{macrocode}
+\def\@currentHref{}
+\def\refstepcounter#1{\stepcounter{#1}%
+    \edef\@currentcounter{#1}%
+    \xdef\@currentHref {#1.\csname theH#1\endcsname}%
+    \protected@edef\@currentlabel
+       {\csname p@#1\expandafter\endcsname\csname the#1\endcsname}%
+}
+%    \end{macrocode}
+%
+% \subsection{Assigning the headings level numbers}
+% [kernel?]
+%
```

Comment:
once that heading stuff is turned into template instances the template should set that number. (should be a note for the future)